ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1298,9 +1306,189 @@ Set<StoreKey> findMissingKeys(List<StoreKey> keys) throws StoreException {
+            segment.getStartOffset());
+        cachedKeys.addFirst(new Pair<>(entry.getKey(), allStoreKeys));
+        if (cachedKeys.size() > config.storeCacheSizeForFindMissingKeysInBatchMode) {
+          cachedKeys.removeLast();
+        }
+      } else {
+        logger.trace("Index {}: In findMissingKeysInbBatch, hitting unsealed index segment, stop using cache : {}",
+            dataDir, segment.getStartOffset());
+        // We are hitting unsealed index segment, there is no need to cache keys anymore
+        shouldUseCachedKeys = false;
+      }
+    } else {
+      missingKeys.add(key);
+    }
+    return shouldUseCachedKeys;
+  }
+
+  /**
+   * Find the input key by searching it in latest cached index segment's next segment. If the next segment exists and it
+   * contains the input key, then add the next segment to the cache. Otherwise, use {@link #findKey} to find the input
+   * key.
+   * @param key The input key
+   * @param missingKeys Set of missing keys. If we can't find the input key, add the input key to this set.
+   * @return A boolean value to indicate if we should keep use cache anymore. False means clear cache and don't use it again.
+   * @throws StoreException
+   */
+  private boolean findKeyInNextSegment(StoreKey key, Set<StoreKey> missingKeys) throws StoreException {
+    boolean shouldUseCachedKeys = true;
+    Offset latestIndexSegmentOffset = cachedKeys.get(0).getFirst();
+    Map.Entry<Offset, IndexSegment> entryAfter = validIndexSegments.higherEntry(latestIndexSegmentOffset);
+    if (entryAfter == null) {
+      // Sanity checking, this shouldn't happen
+      // We only cache sealed index segment, which means all cached index segment can't be the last one
+      logger.error("Index {}: In findMissingKeysInbBatch, searching for next index segment returns null for offset: {}",
+          dataDir, latestIndexSegmentOffset);
+      if (findKey(key) == null) {
+        missingKeys.add(key);
+      }
+    } else {
+      if (cachedKeys.stream().map(Pair::getFirst).anyMatch(offset -> offset.equals(entryAfter.getKey()))) {
+        // if next index segment already in the cache, don't bother to search key in next index segment
+        return findKeyAndAddIndexSegmentToCache(key, missingKeys);
+      }
+      IndexSegment nextSegment = entryAfter.getValue();
+      NavigableSet<IndexValue> values = nextSegment.find(key);
+      if (values == null || values.isEmpty()) {
+        shouldUseCachedKeys = findKeyAndAddIndexSegmentToCache(key, missingKeys);
+      } else {
+        if (nextSegment.isSealed()) {
+          Set<StoreKey> allStoreKeys = nextSegment.getAllStoreKeys();
+          // Adding this index segment to the front of the list
+          logger.info("Index {}: In findMissingKeyInbBatch, adding next segment : {}", dataDir,
+              nextSegment.getStartOffset());
+          cachedKeys.addFirst(new Pair<>(nextSegment.getStartOffset(), allStoreKeys));
```

Comment:
QQ: one thing confused me a little bit is, if we can find the key in next index segment, we add it to the cache, but we are not sure if the next segment has the final state of the key. But findKeyAndAddIndexSegmentToCache will always find the key's final state and add corresponding index segment. So can you explain a little bit about when do we need to find the final state?

Reply:
The reason why we are using findKey is not to find the final state, it's just a tool to find any segment that contains this key. The ultimate goal is to check if this key exists or not, we don't really care about the final state. But I don't have better tool to use than just calling findKey to the keys missing from the cache.